File: 704 Binary Search.cpp
class Solution {
public:
    // 左闭右开写法
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size();// 注意这里不需要-1
        while (left < right) {// 左闭右开区间
            int mid = left + ((right - left) >> 1); // 注意，这里的右移符号一定要括号起来
            if (nums[mid] < target)
                left = mid + 1; // 由于左闭右开，mid不可能是target，所以一定要mid + 1
            else if (nums[mid] > target)
                right = mid;
            else
                return mid;
        }
        return -1;
    }
};

File: 704 Binary Search.py
def search(nums, target) -> int:
    mid = len(nums) // 2
    left = 0
    right = len(nums)
    print(f"mid: {mid}, left: {left}, right: {right}")
    while(left < right):
        if nums[mid] == target:
            return mid
        elif nums[mid] > target:
            right = mid
            mid = mid // 2
            print(f"mid: {mid}, left: {left}, right: {right}")
        elif nums[mid] < target:
            left = mid + 1
            mid = mid + (right - mid) // 2
            print(f"mid: {mid}, left: {left}, right: {right}")
    return -1

nums = [2,5]
print(search(nums, 2))

# 这题主要的问题在于left = mid + 1，使用的是左闭右开的区间
# 而且不要用递归，直接在函数内进行while循环